geode-core/src/main/java/org/apache/geode/internal/tcp/Connection.java

```diff
@@ -1947,14 +1947,14 @@ void releaseMsgDestreamer(short msgId, MsgDestreamer md) {
   }
 
   private void sendFailureReply(int rpId, String exMsg, Throwable ex, boolean directAck) {
-    ReplySender dm = null;
+    ReplyException exception = new ReplyException(exMsg, ex);
     if (directAck) {
-      dm = new DirectReplySender(this);
+      ReplySender dm = new DirectReplySender(this);
+      ReplyMessage.send(getRemoteAddress(), rpId, exception, dm);
     } else if (rpId != 0) {
-      dm = this.owner.getDM();
-    }
-    if (dm != null) {
-      ReplyMessage.send(getRemoteAddress(), rpId, new ReplyException(exMsg, ex), dm);
+      DistributionManager dm = this.owner.getDM();
```

Comment:
Functionally this will work but have you considered using `owner.getMembershipManager().send()` to send the reply message?  That will bypass the lock that you're currently hitting with dm.putOutgoing().  You'd need a new method in ReplyMessage to create one w/o sending it, I suppose.
I think that would be more in-line with what this code is doing with DirectReplySenders.  In that case we're not blocking on the waitUntilReadyToSendMessages lock but are immediately sending a reply.